### PR TITLE
"Trust" privileges for Applications

### PR DESCRIPTION
**Pull request by ExternalReality**
## Description of change

This change adds a command to allow users to authorize an application cloud/model access credentials. It is a shorthand for altering the application configuration of the application.

## QA steps

- Enable CAAS feature flag and try out the commands
```bash
JUJU_DEV_FEATURE_FLAGS=caas <command>
```

## Documentation changes

Yes, document the new commands (once the feature flag is removed)

## Bug reference

N/A

### CODE REVIEW

File: apiserver/facades/client/application/privileges.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package application
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/environschema.v1"
+)
+
+const privilegeConfigOptionName = "elevated_privileges"
```
Comment (jameinel) [inline apiserver/facades/client/application/privileges.go]: I believe most config strings are "-" separated. Just like "juju bootstrap --config http-proxy" this should be "elevated-privileges".

File: apiserver/facades/client/application/privileges.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	privilegeConfigOptionName: {
+		Description: "blah blah",
+		Type:        environschema.Tbool,
+		Group:       environschema.JujuGroup,
+	},
+}
+
+//PrivilegeSchema builder pattern which returns an existing schema augmented
+//with the application privilege schema fields.
+func PrivilegeSchema(providerFields environschema.Fields) (environschema.Fields, error) {
```
Comment (jameinel) [inline apiserver/facades/client/application/privileges.go]: Why have PrivilegeSchema and addPrivilegeSchema where one only exists to call the other one.

Is there an intent to do more than just addPr... here?
If not, then I would just say move the code of addPrivilege into this function, and update callers to call the single function.

File: cmd/juju/application/elevateprivileges.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	c.ModelCommandBase.SetFlags(f)
+}
+
+// Init is part of the cmd.Command interface.
+func (c *elevatePrivilegesCommand) Init(args []string) error {
+	if len(args) == 0 {
+		return errors.New("no application name specified")
+	}
+	c.applicationName = args[0]
+	return c.parseSet([]string{"elevated_privileges=true"})
```
Comment (jameinel) [inline cmd/juju/application/elevateprivileges.go]: s/_/-/

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -317,6 +318,11 @@ type DeployCommand struct {
 	// NewAPIRoot stores a function which returns a new API root.
 	NewAPIRoot func() (DeployAPI, error)
 
+	// Trust signifies that the charm should be deployed with access to
+	// trusted credentials. That is, hooks run by the charm can access
+	// cloud credentials and other trusted access credentials.
+	Trust bool
```
Comment (jameinel) [inline cmd/juju/application/deploy.go]: I'm a little concerned that the flag you pass is:
 juju deploy --trust
but that actually links to
 juju elevate-privileges
and
 juju config elevated-privileges

is "juju deploy --elevate-privileges" or "juju deploy --with-elevated-privileges" is overly verbose?

File: apiserver/facades/client/application/privileges.go (hunk 1 above)
Comment (ExternalReality) [reply apiserver/facades/client/application/privileges.go]: The name is going to be changed to "trust" for both the option and flag. This is just a spike to show how implementing on top of config might look.

File: cmd/juju/application/deploy.go [hunk 5]
```diff
@@ -767,6 +774,11 @@ func (c *DeployCommand) deployCharm(
 		appConfig[k] = v.(string)
 	}
 
+	// Expand the trust flag into the appConfig
+	if c.Trust {
+		appConfig[app.TrustConfigOptionName] = strconv.FormatBool(c.Trust)
+	}
```
Comment (ExternalReality) [inline cmd/juju/application/deploy.go]: If the flag is present we just splice the value into the configuration parameters. We try to reuse as much of the configuration infrastructure as possible.

**Comment by ExternalReality** [comment]
!!build!!

**Comment by ExternalReality** [comment]
!!build!!

**Comment by ExternalReality** [comment]
!!build!!

File: apiserver/facades/client/application/get_test.go [hunk 6]
```diff
@@ -112,8 +113,15 @@ func (s *getSuite) TestClientApplicationGetIAASModelSmoketest(c *gc.C) {
 				"value":       "My Title",
 			},
 		},
-		ApplicationConfig: map[string]interface{}{},
-		Series:            "quantal",
+		ApplicationConfig: map[string]interface{}{
+			"trust": map[string]interface{}{
```
Comment (manadart) [inline apiserver/facades/client/application/get_test.go]: I favour "trust" as a verb and "trusted" as an adjective.

File: apiserver/facades/client/application/trust.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
+package application
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/schema"
+	"gopkg.in/juju/environschema.v1"
+)
+
+// TrustConfigOptionName is the option name used to set trust level in application configuration.
+const TrustConfigOptionName = "trust"
```
Comment (manadart) [inline apiserver/facades/client/application/trust.go]: ... so this would be "trusted".

**Comment by ExternalReality** [comment]
$$MERGE$$

**Pull request merged**